### PR TITLE
unionfs-fuse: enable darwin

### PR DESCRIPTION
**Pull request by kalekseev**
###### Motivation for this change

Provide unionfs on macOS.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 115318` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>unionfs-fuse</li>
  </ul>
</details>

File: pkgs/tools/filesystems/unionfs-fuse/default.nix [hunk 1]
```diff
@@ -39,7 +47,7 @@ stdenv.mkDerivation rec {
     description = "FUSE UnionFS implementation";
     homepage = "https://github.com/rpodgorny/unionfs-fuse";
     license = licenses.bsd3;
-    platforms = platforms.linux;
+    platforms = platforms.linux ++ platforms.darwin;
```
Comment (SuperSandro2000) [inline pkgs/tools/filesystems/unionfs-fuse/default.nix:50]:
```suggestion
    platforms = platforms.unix;
```

File: pkgs/tools/filesystems/unionfs-fuse/default.nix [hunk 2]
```diff
@@ -12,23 +12,31 @@ stdenv.mkDerivation rec {
   };
 
   patches =
```
Comment (SuperSandro2000) [inline pkgs/tools/filesystems/unionfs-fuse/default.nix:14]:
```suggestion
  patches = [
```

File: pkgs/tools/filesystems/unionfs-fuse/default.nix [hunk 3]
```diff
@@ -12,23 +12,31 @@ stdenv.mkDerivation rec {
...
-    [ # Prevent the unionfs daemon from being killed during
+    [
+      # Prevent the unionfs daemon from being killed during
       # shutdown. See
       # http://www.freedesktop.org/wiki/Software/systemd/RootStorageDaemons/
       # for details.
       ./prevent-kill-on-shutdown.patch
     ];
```
Comment (SuperSandro2000) [inline pkgs/tools/filesystems/unionfs-fuse/default.nix:21]:
```suggestion
    # Prevent the unionfs daemon from being killed during
    # shutdown. See
    # http://www.freedesktop.org/wiki/Software/systemd/RootStorageDaemons/
    # for details.
    ./prevent-kill-on-shutdown.patch
  ];
```

File: pkgs/tools/filesystems/unionfs-fuse/default.nix [hunk 4]
```diff
@@ -12,23 +12,31 @@ stdenv.mkDerivation rec {
...
+  postPatch = lib.optionals stdenv.isDarwin ''
+    substituteInPlace CMakeLists.txt \
+      --replace '/usr/local/include/osxfuse/fuse' '${osxfuse}/include/osxfuse/fuse'
+    substituteInPlace src/CMakeLists.txt \
+      --replace 'target_link_libraries(unionfs fuse pthread)' 'target_link_libraries(unionfs osxfuse pthread)'
+  '';
+
   nativeBuildInputs = [ cmake ];
-  buildInputs = [ fuse ];
+  buildInputs = lib.optional stdenv.isLinux fuse ++ lib.optional stdenv.isDarwin osxfuse;
```
Comment (Mic92) [inline pkgs/tools/filesystems/unionfs-fuse/default.nix:31]: I think everyone else uses fuse:

```suggestion
  buildInputs = [ (if stdenv.isDarwin osxfuse else fuse) ];
```

File: pkgs/tools/filesystems/unionfs-fuse/default.nix (hunk 1 above)
Comment (kalekseev) [reply pkgs/tools/filesystems/unionfs-fuse/default.nix:50]: If I understand it right that would enable all unixes while that pr only fixes darwin build. https://github.com/NixOS/nixpkgs/blob/4d100429b57ea7406576c736448e9921ac4bdffc/lib/systems/inspect.nix#L38  Do we have some agreement that platform.unix is basically linux + darwin?

Comment (SuperSandro2000) [reply pkgs/tools/filesystems/unionfs-fuse/default.nix]: > Do we have some agreement that platform.unix is basically linux + darwin?

Basically it is. platforms is mostly for hydra and we do not have any other OS on hydra and BSD, SunOS, Cygwin and Redox is mostly untested all over nixpkgs.

File: pkgs/tools/filesystems/unionfs-fuse/default.nix [hunk 5]
```diff
@@ -11,24 +11,31 @@ stdenv.mkDerivation rec {
...
+  buildInputs = [ (if stdenv.isDarwin then osxfuse else fuse) ];
 
   # Put the unionfs mount helper in place as mount.unionfs-fuse. This makes it
   # possible to do:
   #   mount -t unionfs-fuse none /dest -o dirs=/source1=RW,/source2=RO
   #
   # This must be done in preConfigure because the build process removes
   # helper from the source directory during the build.
-  preConfigure = ''
+  preConfigure = lib.optionals (!stdenv.isDarwin) ''
```
Comment (SuperSandro2000) [inline pkgs/tools/filesystems/unionfs-fuse/default.nix:38]:
```suggestion
  preConfigure = lib.optionalString (!stdenv.isDarwin) ''
```

**Pull request merged**